libsolidity/experimental/analysis/TypeInference.cpp

```diff
@@ -0,0 +1,1221 @@
+	Type functionType = TypeSystemHelpers{m_typeSystem}.functionType(argumentsType, returnType);
+
+	m_currentFunctionType = functionType;
+	functionAnnotation.type = functionType;
+
+
+	_functionDefinition.parameterList().accept(*this);
+	unify(argumentsType, typeAnnotation(_functionDefinition.parameterList()), _functionDefinition.parameterList().location());
+	if (_functionDefinition.experimentalReturnExpression())
+	{
+		ScopedSaveAndRestore expressionContext{m_expressionContext, ExpressionContext::Type};
+		_functionDefinition.experimentalReturnExpression()->accept(*this);
+		unify(
+			returnType,
+			typeAnnotation(*_functionDefinition.experimentalReturnExpression()),
+			_functionDefinition.experimentalReturnExpression()->location()
+		);
+	}
+	else
+		unify(returnType, m_unitType, _functionDefinition.location());
+
+	if (_functionDefinition.isImplemented())
+		_functionDefinition.body().accept(*this);
+
+	return false;
+}
+
+void TypeInference::endVisit(Return const& _return)
+{
+	solAssert(m_currentFunctionType);
+	Type functionReturnType = std::get<1>(TypeSystemHelpers{m_typeSystem}.destFunctionType(*m_currentFunctionType));
+	if (_return.expression())
+		unify(functionReturnType, typeAnnotation(*_return.expression()), _return.location());
+	else
+		unify(functionReturnType, m_unitType, _return.location());
+}
+
+void TypeInference::endVisit(ParameterList const& _parameterList)
+{
+	auto& listAnnotation = annotation(_parameterList);
+	solAssert(!listAnnotation.type);
+	listAnnotation.type = TypeSystemHelpers{m_typeSystem}.tupleType(
+		_parameterList.parameters() | ranges::views::transform([&](auto _arg) { return typeAnnotation(*_arg); }) | ranges::to<std::vector<Type>>
+	);
+}
+
+bool TypeInference::visit(TypeClassDefinition const& _typeClassDefinition)
+{
+	solAssert(m_expressionContext == ExpressionContext::Term);
+	auto& typeClassDefinitionAnnotation = annotation(_typeClassDefinition);
+	if (typeClassDefinitionAnnotation.type)
+		return false;
+
+	typeClassDefinitionAnnotation.type = type(&_typeClassDefinition, {});
+
+	{
+		ScopedSaveAndRestore expressionContext{m_expressionContext, ExpressionContext::Type};
+		_typeClassDefinition.typeVariable().accept(*this);
+	}
+
+	std::map<std::string, Type> functionTypes;
+
+	solAssert(m_analysis.annotation<TypeClassRegistration>(_typeClassDefinition).typeClass.has_value());
+	TypeClass typeClass = m_analysis.annotation<TypeClassRegistration>(_typeClassDefinition).typeClass.value();
+	TypeVariable typeVar = m_typeSystem.typeClassVariable(typeClass);
+	auto& typeMembersAnnotation = annotation().members[typeConstructor(&_typeClassDefinition)];
+
+	for (auto subNode: _typeClassDefinition.subNodes())
+	{
+		subNode->accept(*this);
+		auto const* functionDefinition = dynamic_cast<FunctionDefinition const*>(subNode.get());
+		solAssert(functionDefinition);
+		// TODO: need polymorphicInstance?
+		auto functionType = polymorphicInstance(typeAnnotation(*functionDefinition));
```

Comment:
I think we do not. This type is used in multiple places and needs to be generalized separately in each one. So this seems completely superfluous and if it actually affect behavior then I'd say it must be due to missing `polymorphicInstance()` at the point where the value is used.

Reply:
Yeah, the gist of the comment was "likely we don't need it and should remove it" without double-checking right away :-).

Reply:
#14635 now removes this.